pkg/operator/controller/gateway-labeler/controller.go

```diff
@@ -0,0 +1,222 @@
+	// Create a new cache for gateways so it can watch all namespaces.
+	// (Using the operator cache for gateways in all namespaces would cause
+	// it to cache other resources in all namespaces.)
+	gatewaysCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Make sure the manager starts the cache with the other runnables.
+	mgr.Add(gatewaysCache)
+
+	reconciler := &reconciler{
+		cache:    gatewaysCache,
+		client:   mgr.GetClient(),
+		recorder: mgr.GetEventRecorderFor(controllerName),
+	}
+	c, err := controller.NewUnmanaged(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+
+	// Watch gatewayclasses so that when one is created with our controller
+	// name or is updated to change its controller name to our controller
+	// name, then we update labels on gateways that reference that
+	// gatewayclass.
+	key, value := operatorcontroller.IstioRevLabelKey, operatorcontroller.IstioName("").Name
+	gatewayClassToGateways := func(ctx context.Context, o client.Object) []reconcile.Request {
+		requests := []reconcile.Request{}
+
+		// Using an index might be more efficient than listing all
+		// gateways and then filtering.  However, setting up an index is
+		// complicated by the fact that the Gateway API CRDs might not
+		// exist when this controller is initialized.
+
+		var gateways gatewayapiv1.GatewayList
+		if err := reconciler.cache.List(ctx, &gateways); err != nil {
+			log.Error(err, "Failed to list gateways for gatewayclass", "gatewayclass", o.GetName())
+			return requests
+		}
+
+		for i := range gateways.Items {
+			if gateways.Items[i].Labels[key] == value {
+				continue
+			}
+			if string(gateways.Items[i].Spec.GatewayClassName) != o.GetName() {
+				continue
+			}
+			request := reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Namespace: gateways.Items[i].Namespace,
+					Name:      gateways.Items[i].Name,
+				},
+			}
+			requests = append(requests, request)
+		}
+
+		return requests
+	}
+	gatewayClassPredicate := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			gatewayClass := e.Object.(*gatewayapiv1.GatewayClass)
+
+			if gatewayClass.Spec.ControllerName == operatorcontroller.OpenShiftGatewayClassControllerName {
+				return true
+			}
+
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// Don't reconcile gateways when the gatewayclass is
+			// deleted.  If Istio created resources for a gateway,
+			// then we want Istio to continue managing those
+			// resources.  The user can remove the label or delete
+			// the gateway and related resources if desired.
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			old := e.ObjectOld.(*gatewayapiv1.GatewayClass)
+			new := e.ObjectNew.(*gatewayapiv1.GatewayClass)
+
+			switch {
+			case old.Spec.ControllerName == new.Spec.ControllerName:
```

Comment:
ControllerName is immutable so we could simplify this.  https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1/gatewayclass_types.go#L81-L86

Reply:
That's an interesting point.  I wonder whether there could be any edge cases where controller-runtime could register delete and recreate as an update?  Probably not, and this logic is probably not necessary, but it feels more correct to handle updates just in case something really weird happens.